File: problems/CSES/Dynamic_Programming/two_sets_ii/two_sets_ii.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    const int MOD = 1000000007;
    int n;
    cin >> n;

    ll total_sum = (ll)n * (n + 1) / 2;

    if (total_sum % 2 != 0) {
        cout << 0 << "\n";
        return 0;
    }

    ll target = total_sum / 2;

    vector<ll> dp(target + 1, 0);
    dp[0] = 1;

    for (int x = 1; x <= n; x++) {
        for (ll curr_sum = target; curr_sum >= x; curr_sum--) {
            dp[curr_sum] += dp[curr_sum - x];

            if (dp[curr_sum] >= MOD) {
                dp[curr_sum] -= MOD;
            }
        }
    }

    cout << dp[target] * 500000004 % MOD << "\n";

    return 0;
}

File: problems/CSES/Dynamic_Programming/two_sets_ii/two_sets_ii.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    MOD = 10**9 + 7
    n = int(input())

    total_sum = n * (n + 1) // 2

    if (total_sum % 2 != 0):
        print(0)
        return

    target = total_sum // 2

    dp = [0] * (target + 1)
    dp[0] = 1

    for x in range(1, n):
        for curr_sum in range(target, x - 1, -1):
            dp[curr_sum] += dp[curr_sum - x]

            if (dp[curr_sum] >= MOD):
                dp[curr_sum] -= MOD

    print(dp[target])

main()
